src/com/amazon/ion/impl/bin/utf8/Pool.java

```diff
@@ -59,6 +69,17 @@ public T getOrCreate() {
     public void returnToPool(T object) {
         // The `offer` method does not block. If the queue is full, it returns `false` immediately.
         // If the provided instance cannot be added to the pool, we discard it silently.
-        bufferQueue.offer(object);
+        if (size.getAndIncrement() < MAX_QUEUE_SIZE) {
+            bufferQueue.offer(object);
+        } else {
+            // The pool was full. Since the size was optimistically incremented, decrement it now.
+            // Note: there is a race condition here that is deliberately allowed as an optimization.
+            // Under high contention, multiple threads could end up here before the first one
+            // decrements the size, causing blocks to be dropped wastefully. This is not harmful
+            // because blocks will be re-allocated when necessary; the pool is kept as close as
```

Comment:
```suggestion
            // Under high contention, multiple threads could end up here before the first one
            // decrements the size, causing objects to be dropped wastefully. This is not harmful
            // because objects will be re-allocated when necessary; the pool is kept as close as
```

Reply:
Done.